File: app/src/main/java/game/ivan/ashancalculator/tags/presenter/TagsPresenter.java
package game.ivan.ashancalculator.tags.presenter;

import com.hannesdorfmann.mosby.mvp.MvpBasePresenter;

import javax.inject.Inject;

import game.ivan.ashancalculator.AshanApplication;
import game.ivan.ashancalculator.database.DatabaseTagsManager;
import game.ivan.ashancalculator.database.models.Tags;
import game.ivan.ashancalculator.tags.presenter.dagger.TagsPresenterComponent;
import game.ivan.ashancalculator.tags.view.TagsView;

/**
 * Created by ivan on 20.12.16.
 */

public class TagsPresenter extends MvpBasePresenter<TagsView> {

    public static final int DEFAULT_DIVISION_VALUE = 1;

    @Inject
    DatabaseTagsManager databaseManager;
    TagsPresenterComponent component;


    public TagsPresenter(){
        component = AshanApplication.getComponent().createTagsPresenterComponent();
        component.injectTagsPresenter(this);
    }

    public void addTag(Tags tag){
        databaseManager.insertTag(tag);
        loadTags(false);
    }

    public void loadTags(boolean lock){
        if(!lock && isViewAttached()){
            getView().refreshList(databaseManager.readAllRecord());
        }
    }

    public void deleteTag(Tags tag){
        databaseManager.deleteTag(tag);
        loadTags(false);
    }

    public void detachView(boolean retainPresenterInstance){
        super.detachView(retainPresenterInstance);
    }

    public int checkDivision(String division){
        try {
            return Integer.valueOf(division);
        } catch(NumberFormatException | NullPointerException e){
            return DEFAULT_DIVISION_VALUE;
        }

    }

}

File: app/src/main/java/game/ivan/ashancalculator/tags/presenter/TagsPresenter.kt
package game.ivan.ashancalculator.tags.presenter

import com.hannesdorfmann.mosby.mvp.MvpBasePresenter
import game.ivan.ashancalculator.AshanApplication
import game.ivan.ashancalculator.database.DatabaseTagsManager
import game.ivan.ashancalculator.database.models.Tags
import game.ivan.ashancalculator.tags.view.TagsView

/**
 * Created by ivan on 05.02.2017.
 */
class TagsPresenter(): MvpBasePresenter<TagsView>(){

    lateinit var databaseManager: DatabaseTagsManager

    fun addTag(tag: Tags){
        databaseManager.insertTag(tag)
        loadTags(false)
    }

    fun loadTags(lock:Boolean){
        if(!lock && isViewAttached){
            view!!.refreshList(databaseManager.readAllRecord())
        }
    }

    fun deleteTag(tag:Tags){
        databaseManager.delteTag(tag)
        loadTags(false)
    }

    override fun attachView(view: TagsView?) {
        super.attachView(view)
        databaseManager = DatabaseTagsManager(AshanApplication.instance)
    }

    override fun detachView(retainInstance: Boolean) {
        super.detachView(retainInstance)
        databaseManager.destroy()
    }
}